### DISCUSSION TITLE
Is it safe to call OPENSSL_thread_stop multiple times?

### DISCUSSION THREAD

**Discussion by jay** [openssl/openssl#22831] [Q&A] [answered]
A libcurl dll built with statically linked OpenSSL may cause a memory leak because OPENSSL_thread_stop is not called. We could call OPENSSL_thread_stop in our DllMain on DLL_THREAD_DETACH, however if OpenSSL is a DLL then it also calls that function in its DllMain on DLL_THREAD_DETACH. Is it allowed to call the function twice?

I do not see it documented and I can't tell from the source. It looks like the function tries to retrieve a tls local value and passes it on to init_thread_remove_handlers even if that value is NULL.

**Comment by mattcaswell** [comment, marked as answer]
Yes, I believe this should be fine.

`OPENSSL_thread_stop` calls `init_get_thread_local` to obtain the THREAD_EVENT_HANDLER list. On a second call of `OPENSSL_thread_stop` this return value should be NULL.

It then calls `init_thread_stop` with that value - which just immediately returns if the handler list is NULL, and `init_thread_remove_handlers` which is a no-op if the handler list is NULL.

Finally it calls `OPENSSL_free` with the NULL pointer - which is also fine and a no-op.

**Reply by jay** [reply]
Thanks. curl/curl#12408

**Comment by BrianInglis** [comment]
Local patch applied by Cygwin OpenSSL maintainer to test release; see [Test: openssl 3.0.16-2, mingw64-{i686,x86_64}-openssl-3.0.16-2](https://cygwin.com/pipermail/cygwin-announce/2025-May/012341.html) and 
[openssl and libcurl issue with cargo](https://cygwin.com/pipermail/cygwin/2025-May/258203.html)
